File: WOT-iOS/WOT-iOS/core/adapters/instances/Chassis/WOTWebResponseAdapterChassis.m
//
//  WOTWebResponseAdapterChassis.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 6/23/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTWebResponseAdapterChassis.h"
#import "Tankchassis.h"

@implementation WOTWebResponseAdapterChassis

- (void)parseData:(id)data error:(NSError *)error {
    
    if (error) {
        
        NSLog(@"%@",error.localizedDescription);
        return;
    }
    
    NSDictionary *tankChassesDictionary = data[WOT_KEY_DATA];
    
    NSArray *tankChassisArray = [tankChassesDictionary allKeys];
    
    NSManagedObjectContext *context = [[WOTCoreDataProvider sharedInstance] workManagedObjectContext];
    for (NSString *key in tankChassisArray) {
        
        NSDictionary *tankChassisJSON = tankChassesDictionary[key];
        
        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@",WOT_KEY_MODULE_ID,tankChassisJSON[WOT_KEY_MODULE_ID]];
        Tankchassis *tankChasses = [Tankchassis findOrCreateObjectWithPredicate:predicate inManagedObjectContext:context];
        [tankChasses fillPropertiesFromDictionary:tankChassisJSON];
    }
    
    if ([context hasChanges]) {
        
        NSError *error = nil;
        [context save:&error];
    }
}

@end

File: WOT-iOS/WOTData/WOTData/CoreData/model/obsolette/jsonparsing/Chassis/WOTWebResponseAdapterChassis.swift
//
//  WOTWebResponseAdapterChassis.swift
//  WOTData
//
//  Created by Pavel Yeshchyk on 1/15/20.
//  Copyright © 2020 Pavel Yeshchyk. All rights reserved.
//

import Foundation

@objc
open class WOTWebResponseAdapterChassis: NSObject, WOTWebResponseAdapter {
    
    public func parseData(_ binary: Data?, error: Error?, nestedRequestsCallback: JSONMappingNestedRequestsCallback?) -> Error? {
        
        return binary?.parseAsJSON { (json) in

            guard let keys = json?.keys, keys.count != 0 else {
                return
            }
            
            let context = WOTTankCoreDataProvider.sharedInstance.workManagedObjectContext
            context.perform {
                keys.forEach { (key) in
                    guard let chassis = json?[key] as? [AnyHashable: Any] else { return }
                    guard let module_id = chassis[WGJsonFields.module_id] as? String else { return }
                    let predicate = NSPredicate(format: "%K == %@", #keyPath(Tankchassis.module_id), module_id)
                    guard let newObject = NSManagedObject.findOrCreateObject(forClass:Tankchassis.self, predicate: predicate, context: context) as? Tankchassis else { return }
                    
                    newObject.mapping(fromJSON: chassis, into: context, completion: nestedRequestsCallback)
                }
                
                if context.hasChanges {
                    do {
                        try context.save()
                    } catch {
                        print("\(#file) \(#function) at \(error.localizedDescription)")
                    }
                }
            }
        }
    }
}
